src/main/java/org/prebid/server/auction/AmpRequestFactory.java

```diff
@@ -123,6 +142,138 @@ private static BidRequest fillExplicitParameters(BidRequest bidRequest, RoutingC
+            formats = new ArrayList<>();
+
+            if (format != null) {
+                formats.add(format);
+            }
+
+            if (CollectionUtils.isNotEmpty(multiSizeFormats) && ow == null && oh == null) {
+                formats.addAll(multiSizeFormats);
+            }
+        }
+        return formats;
+    }
+
+    private static Site overrideSitePage(Site site, HttpServerRequest request) {
+        final String canonicalUrl = canonicalUrl(request);
+
+        if (StringUtils.isBlank(canonicalUrl)) {
+            return site;
+        }
+
+        final Site.SiteBuilder siteBuilder = site == null ? Site.builder() : site.toBuilder();
+        return siteBuilder.page(canonicalUrl).build();
+    }
+
+    private Imp overrideImp(Imp imp, HttpServerRequest request) {
+        final String tagId = request.getParam(SLOT_REQUEST_PARAM);
+        final List<Format> overwrittenFormats = overrideBannerFormats(request);
+        if (StringUtils.isNotBlank(tagId) || CollectionUtils.isNotEmpty(overwrittenFormats)) {
+            return imp.toBuilder()
+                    .tagid(StringUtils.isNotBlank(tagId) ? tagId : imp.getTagid())
+                    .banner(overrideBanner(imp.getBanner(), overwrittenFormats))
+                    .build();
+        }
+        return null;
+    }
+
+    private static Banner overrideBanner(Banner banner, List<Format> formats) {
+        return banner != null && CollectionUtils.isNotEmpty(formats)
+                ? banner.toBuilder().format(formats).build()
+                : banner;
+    }
+
+    private Long updateTimeout(HttpServerRequest request) {
+        Long timeout;
+        try {
+            timeout = Long.parseLong(request.getParam(TIMEOUT_REQUEST_PARAM));
+        } catch (NumberFormatException e) {
+            timeout = null;
+        }
+        return timeout != null ? timeout - timeoutAdjustmentMs : null;
+    }
+
+    private static BidRequest updateBidRequest(BidRequest bidRequest, Site outgoingSite, Imp outgoingImp,
+                                               Long timeout) {
+        final boolean isValidTimeout = timeout != null && timeout > 0;
+        if (outgoingSite != null || outgoingImp != null || isValidTimeout) {
+            return bidRequest.toBuilder()
+                    .site(outgoingSite != null ? outgoingSite : bidRequest.getSite())
+                    .imp(outgoingImp != null ? Collections.singletonList(outgoingImp) : bidRequest.getImp())
+                    .tmax(isValidTimeout ? timeout : bidRequest.getTmax())
+                    .build();
+        }
+        return bidRequest;
+    }
+
+    private static Integer parseIntParam(HttpServerRequest request, String name) {
+        final String param = request.getParam(name);
+        return parseInt(param);
+    }
+
+    private static Integer parseInt(String param) {
+        try {
+            return Integer.parseInt(param);
+        } catch (NumberFormatException e) {
+            return null;
```

Comment:
having doubts if is it worth to be logged?

Reply:
If that parameter matters for auction it would be correct to reject such request asking to fix it

Reply:
all parameters are optional.